src/core_plugins/kibana/public/discover/controllers/discover.js

```diff
@@ -292,6 +293,8 @@ function discoverController(
     $state.save();
   });
 
+  $scope.$watch('state.isVisualizationVisible', $state.save.bind($state));
```

Comment:
Couldn't this go in the `showVis` and `hideVis` methods? Then we wouldn't have to create an additional watch.